merge interfaces that have an optional scalartype parameter

nairbv opened the pull request:
This change is backwards incompatible in *C++ only* on mean(), sum(), and prod() interfaces that accepted either of:
```
Tensor sum(IntArrayRef dim, bool keepdim=false) const;	
Tensor sum(IntArrayRef dim, ScalarType dtype) const;
```
but now to specify both the dim and dtype will require the keepdim parameter:
```
Tensor sum(IntArrayRef dim, bool keepdim=false, c10::optional<ScalarType> dtype=c10::nullopt) const;
```

[xla ci]

nairbv committed:
merge interfaces that have an optional scalartype parameter

nairbv committed:
Merge branch 'master' into opt_scalartype_interfaces

nairbv commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

nairbv commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

nairbv committed:
fix flake8 issues

nairbv committed:
fix jit tests for optional dtype

nairbv commented:
With latest changes the existing jit tests pass (at least, test_jit.py, when I ran them pre-commit), but it seems to me like I should need to make additional changes to support the dtype arg in `torch/csrc/jit/passes/shape_analysis.cpp`. If that's correct, I'm not sure what the tests to validate those changes would look like.

wanchaol commented on torch/csrc/jit/symbolic_script.cpp:
```diff
@@ -46,23 +46,27 @@ const std::vector<std::string> functions = {
+        def mean_0(self, *, dtype: Optional[int]):
             self_size = self.size()
             self_numel = self.numel()
+            self_scalar_type = self.dtype
```
this does not correctly unwrap the dtype I believe, `self_scalar_type` is still an `Optional[int]`, you will probably need to unwrap it using boolean refinement like `if is not None`

nairbv replied:
I wasn't trying to unwrap, was just trying to be consistent with how the rest of the code is written. e.g. self.scalar_type() gets converted to self_scalar_type (`REPLACEMENTS` in `def saved_variables` in load_derivatives.py).

I think torch.mean() and to() both accept optional. If this is wrong though, how would I write a test that would catch it? It seems like this is working.

wanchaol replied:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L2704 hmmm that's pretty weird, from what i see in the schema, `to` does not take optional dtype, it also does not have default values, so I am not sure why this is working. 

For the mean, it looks like you are changing the API, can you add some tests to test the api in eager and see if Autodiff works in JIT? https://github.com/pytorch/pytorch/blob/master/test/common_methods_invocations.py#L339
The tests right now only have one test case, which does not contains the test case with `dtype`

nairbv replied:
ah, I didn't notice this comment. Added a test of `('mean', (S, S, S), (), 'dtype', (True,), (), (), ident, {'dtype': torch.float64}),`.  Is that what you mean?

passed when I ran locally with:
python test_jit.py TestJitGeneratedAutograd.test_mean_dtype

wanchaol commented on torch/csrc/jit/symbolic_script.cpp:
```diff
@@ -46,23 +46,27 @@ const std::vector<std::string> functions = {
+                   *,
+                   dtype: Optional[int]):
             self_size = self.size()
+            self_scalar_type = self.dtype
```
same here

wanchaol commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -267,7 +267,7 @@ void FunctionParameter::set_default_str(const std::string& str) {
     }
   } else if (type_ == ParameterType::SCALARTYPE) {
-    if (str == "None") {
+    if (str == "None" || str == "c10::nullopt") {
```
why here is `c10::nullopt`? python arg parser should only consume `None` in python rather than `nullopt`

nairbv replied:
hmm... I hadn't thought about it in detail since this did fix an error I was getting at this point in the code. The default value is set in c++ though right? If the C++ function is defined like `Tensor::mean(Tensor self, c10::optional<ScalarType> dtype = c10::nullopt)`, and the python function call is tensor.mean(), where should the parser find the conversion from `c10::nullopt` to `None`?

nairbv commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1146,7 +1146,7 @@ class ShapePropagator {
     //   - has a bool keepdim argument
     static const register_formula_for dim_reduce_ops_with_integer_upcast{
```
In these `register_formula_for` entries it looks like the code is explicitly parsing each argument in the registered function. It seems that since I've added a dtype argument, I also need to parse that. I can guess at what the code should look like to do so, but I don't have a failing tests and I'm not sure how to write one.

nairbv commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1676,11 +1676,11 @@ class ShapePropagator {
       node->output()->setType(tp->withSizesStrides(sizes, tp->strides()));
       return true;
-    } else if (node->matches("aten::sum(Tensor self) -> Tensor")) {
+    } else if (node->matches("aten::sum(Tensor self, *, int? dtype) -> Tensor")) {
```
Similarly here in `PropagateCompleteShapeOnNode`, it looks like this code explicitly handles each argument. I think I need to add similar code to extract the `dtype`, but don't have a failing test and I'm not sure how to write one that will reach this code.

eellison replied:
https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L6726 here is an example of how to write shape propagation tests

nairbv committed:
fix a failing msnpu test

Jun 04 16:33:39 ERROR: test_backwards (__main__.TestMSNPUTensor)
Jun 04 16:33:39 ----------------------------------------------------------------------
Jun 04 16:33:39 Traceback (most recent call last):
Jun 04 16:33:39   File "test_cpp_extensions.py", line 670, in test_backwards
Jun 04 16:33:39     d = c.sum()
Jun 04 16:33:39 RuntimeError: No function registered for schema: sum(Tensor self, ScalarType dtype) -> Tensor

nairbv committed:
attempt to fix msnpu_extension test

nairbv commented on test/cpp_extensions/msnpu_extension.cpp:
```diff
@@ -64,7 +65,7 @@ void init_msnpu_extension() {
   register_extension_backend_op(
     Backend::MSNPU,
-    "sum(Tensor self) -> Tensor", &sum_override);
+    "sum(Tensor self, ScalarType dtype) -> Tensor", &sum_override);
```
Technically this isn't correct since dtype should be optional. That's due to an existing bug, filed here to be fixed separately:

https://github.com/pytorch/pytorch/issues/21416

eellison commented on test/expect/TestScript.test_optional_scalartype.expect:
```diff
@@ -0,0 +1,13 @@
+graph(%a : Tensor,
+      %b : Tensor):
```
Could you find another way to test this other than expect files ? We do not use them. Consider using our filecheck tool as a way of comparing expected textual output

nairbv replied:
removed the expect test and added FileCheck tests

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -720,7 +720,7 @@ class ShapePropagator {
             "aten::log2(Tensor self) -> Tensor",
             "aten::log_sigmoid(Tensor self) -> Tensor",
-            "aten::log_softmax(Tensor self, int dim) -> Tensor",
+            "aten::log_softmax(Tensor self, int dim, *, int? dtype) -> Tensor",
```
The comment above these sets of ops require that the scalar type of the input is preserved. Now that dtype is an argument it is not longer valid to have these ops in this set.

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1044,8 +1044,8 @@ class ShapePropagator {
         {
-            "aten::sum(Tensor self) -> Tensor",
-            "aten::prod(Tensor self) -> Tensor",
+            "aten::sum(Tensor self, *, int? dtype) -> Tensor",
```
Same here

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1146,7 +1146,7 @@ class ShapePropagator {
     static const register_formula_for dim_reduce_ops_with_integer_upcast{
         {
-            "aten::prod(Tensor self, int dim, bool keepdim) -> Tensor",
+            "aten::prod(Tensor self, int dim, bool keepdim, *, int? dtype) -> Tensor",
```
Same here

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1163,7 +1163,7 @@ class ShapePropagator {
         {
             "aten::logsumexp(Tensor self, int[] dim, bool keepdim) -> Tensor",
-            "aten::mean(Tensor self, int[] dim, bool keepdim) -> Tensor",
+            "aten::mean(Tensor self, int[] dim, bool keepdim, *, int? dtype) -> Tensor",
```
here as well

eellison reviewed (commented):
Few comments on the shape propagation. You can check-in https://github.com/pytorch/pytorch/pull/18813 while you're working on it as well (and maybe I should land it).

nairbv committed:
fixing shape analysis for opt-dtypes

nairbv commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1036,45 +1029,84 @@ class ShapePropagator {
+            } else if(upcast_integer && !at::isFloatingType(type->scalarType())) {
+              type = type->toScalarType(at::kLong);
+            }
+            if (type->dim() >= num_reduced_dim && num_reduced_dim > 0) {
```
equivalent previous code was `if (type->dim() > num_reduced_dim)` which seems to have been incorrect / a bug. Wasn't allowing for reductions from 1 to 0 dim.

nairbv commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1146,42 +1178,66 @@ class ShapePropagator {
         [](Node* node) -> type_vec_t {
-          if (auto dim = node->get<std::vector<int64_t>>(attr::dim)) {
-            return multidim_reduce_with_postprocess(
+          if (auto dims = node->namedInput(attr::dim)->node()->inputs().size()) {
```
previous code used `auto dim = node->get<std::vector<int64_t>>(attr::dim)` and then `dim->size()`, which I believe was incorrect and rarely would have given a correct graph for these functions. In my debugging it always returned c10::nullopt. I believe this is because get() from ir.cpp calls `toIValue()` from constants.cpp, which returns `c10::nullopt` whenever `value->node()->kind() != prim::Constant`.

nairbv commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1190,25 +1246,45 @@ class ShapePropagator {
-            return multidim_reduce_with_postprocess(
-                node, /*num_reduced_dim=*/dim->size(), /*upcast_integer=*/true);
+            return reduce_op_handler(
+                node, /*num_reduced_dim=*/num_reduced, /*upcast_integer=*/true, opt_dtype);
```
prior code had upcast=false for mean, but it's not necessary to differentiate since mean isn't implemented for integers anyway.

ailzhang reviewed (commented):
Please run this PR against xla test as well using [xla ci] in the PR body. Thanks!

nairbv commented on test/test_jit.py:
```diff
@@ -12375,6 +12443,14 @@ def fn4(x):
+            return torch.add(a.cumsum(0, dtype=torch.long).sum(dtype=None),
+                             b.cumprod(0, dtype=None).prod(dtype=torch.double))
+
+        self.assertExpectedGraph(fn.graph)
```
forgot to remove the 'expect' based test, removing now

nairbv committed:
remove "expect" based test, was replaced by FileChec()

nairbv committed:
fix a bug in a jit test

nairbv committed:
fix another incorrect guard on type conversion

nairbv committed:
Merge branch 'master' into opt_scalartype_interfaces

nairbv committed:
fix compilation errors caused by merge

nairbv committed:
fix flake8 error

nairbv committed:
[xla ci]

nairbv committed:
fix an issue where onnx segfaults on TestCaffe2BackendEmbed and test_cumsum

nairbv committed:
remove nullopt check when parsing pythong -- cant remember why this was needed

nairbv committed:
fix softmax onnx test

nairbv committed:
fix incorrect indent

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -248,22 +209,29 @@ static Tensor& sum_out(Tensor& result, const Tensor& self, IntArrayRef dim,
 
-Tensor sum(const Tensor &self, ScalarType dtype) {
-  return at::native::sum(self, {}, false, optional<ScalarType>(dtype));
+Tensor sum(const Tensor& self, IntArrayRef dim, bool keepdim) {
```
do we need all these overloaded implementations?  I would expect we only need 3, which is the number of functions specified in native_functions, but there are 5 `sum` functions here.

nairbv replied:
ah not sure why I had these, removed.

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -274,22 +242,23 @@ static Tensor& prod_out(Tensor& result, const Tensor& self, IntArrayRef dim,
 
-Tensor prod(const Tensor &self) {
-  return at::native::prod(self, {}, false, c10::nullopt);
+Tensor& prod_out(Tensor& result, const Tensor& self, int64_t dim, bool keepdim, c10::optional<ScalarType> dtype) {
```
same with prod.

nairbv replied:
for prod it's just a static helper function, I don't see an unused one like there was for sum()

nairbv replied:
err, nevermind, I'll look again, I see you were referencing another file with some of these comments.

nairbv replied:
yeah I think this is correct, looks like your other comments were referencing the BC issue. Here for prod we just have three functions and one static helper.

nairbv committed:
remove unnecessary code found in review

nairbv commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -453,18 +451,14 @@ class CAFFE2_API Tensor {
-  Tensor mean(IntArrayRef dim, bool keepdim=false) const;
-  Tensor mean(IntArrayRef dim, ScalarType dtype) const;
+  Tensor mean(c10::optional<ScalarType> dtype=c10::nullopt) const;
+  Tensor mean(IntArrayRef dim, bool keepdim=false, c10::optional<ScalarType> dtype=c10::nullopt) const;
```
added a comment and the label

nairbv commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1174,27 +1158,15 @@
-# FIXME: These could be combined as optional<ScalarType> but for https://github.com/pytorch/pytorch/issues/6593.
-- func: mean(Tensor self, *, ScalarType dtype) -> Tensor
+- func: mean(Tensor self, *, ScalarType? dtype=None) -> Tensor
+  matches_jit_signature: True
```
removed

nairbv committed:
Merge branch 'master' into opt_scalartype_interfaces

gchanan reviewed (approved):
approving non-JIT changes.

nairbv committed:
Merge branch 'master' into opt_scalartype_interfaces

nairbv committed:
commit re-generated tensormethods after merge conflict

nairbv commented:
@eellison  / @suo / @wanchaol , can one of you approve the jit-related changes here?

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1159,42 +1191,64 @@ class ShapePropagator {
+
+    // Requirements:
+    //   dims           : preserved
+    //   scalar type    : preserved unless specified.
```
`scalar type    : preserved unless specified` is inaccurate since integer_upcast is true

nairbv replied:
right, fixed

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1159,42 +1191,64 @@ class ShapePropagator {
 
     // Requirements:
-    //   dims           : preserved if keepdim == false, 1 smaller otherwise
+    //   dims           : preserved if keepdim == false, dim.size() smaller otherwise
```
is this equivalent to saying `preserved if keepdim == false, 0 otherwise`

nairbv replied:
no because dims comes from the `dim` parameter, not from `self`.  E.g., 3-dimension tensor reduced on two dims gives 1-dim tensor:
```
>>> torch.ones([3,3,3]).sum([0,1])
tensor([9., 9., 9.])
```

eellison commented on test/test_jit.py:
```diff
@@ -6690,6 +6690,73 @@ def test_check_not():
+                    if(not tensor_type.is_floating_point or (dtype is not None and not dtype.is_floating_point)):
+                        if op in ['mean', 'softmax', 'log_softmax']:
+                            continue
+                    return_line = "torch.tensor({}, dtype={}).{}({}dtype={})".format(tensor_data, tensor_type, op, str_args, dtype)
```
Don't some of the ops have different behavior depending on what the inputs tensors dtype is? If so, don't you need to iterate over all pairs of dtypes for the `torch.tensor` dtype arg and the op dtype?

nairbv replied:
I'm not sure what you mean, doesn't it? we nested-loop over dtypes twice, once as `dtype` and once as `tensor_type`.

eellison replied:
oop yea nvm, you're doing it already

eellison commented on test/test_jit.py:
```diff
@@ -6690,6 +6690,73 @@ def test_check_not():
+        self._test_dtype_op_shape(ops, [0, False])
+
+        ops = ['sum', 'mean']
+        self._test_dtype_op_shape(ops, [[0, 1], False], 4)
```
Nit: maybe use kwargs here so it's easier to tell what the arguments mean

nairbv replied:
updated

eellison reviewed (approved):
Looks good to me, thanks for the effort in updating the JIT. I know it's a tricky part of the codebase to navigate.

Please update the test for pairs of dtypes before landing if that comment applies.

nairbv committed:
fix incorrect comment

nairbv committed:
use kwargs

eellison commented:
@ailzhang @wanchaol take a look at autograd ?

nairbv commented:
@pytorchbot retest this please

nairbv commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

nairbv committed:
add a test for mean with dtype to verify jit

nairbv commented on test/common_methods_invocations.py:
```diff
@@ -342,6 +342,7 @@ def method_tests():
         ('mean', (), NO_ARGS, 'scalar', (True,)),
         ('mean', (), (0,), 'scalar_dim', (True,), [0]),
         ('mean', (), (0, True,), 'scalar_keepdim_dim', (True,), [0]),
+        ('mean', (S, S, S), (), 'dtype', (True,), (), (), ident, {'dtype': torch.float64}),
```
adding test with dtype, @wanchaol / @ailzhang

wanchaol reviewed (approved):
looks good to me. Thanks for adding the test! 

Might need to follow up with @ailzhang on the xla test

ailzhang commented:
Yea I will make followup changes on XLA side.

suo commented:
tests are failing master with errors like:
```
Jun 24 15:42:02 	Schema not found for node. File a bug report.
Jun 24 15:42:02 	Node: %21 : Tensor = aten::cumsum(%18, %3, %31) # /opt/conda/lib/python3.6/site-packages/torch/distributions/transforms.py:516:0
```
